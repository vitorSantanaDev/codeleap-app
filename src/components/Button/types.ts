import { ButtonHTMLAttributes, ReactNode } from 'react'

export type ButtonBgColor = 'primary' | 'green' | 'red' | 'white'

export type ButtonProps = {
  children: ReactNode
  minimal?: boolean
  bgColor?: ButtonBgColor
} & ButtonHTMLAttributes<HTMLButtonElement>
