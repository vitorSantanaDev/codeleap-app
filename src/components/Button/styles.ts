import styled, { css } from 'styled-components'
import { ButtonBgColor } from './types'
import { darken, lighten } from 'polished'

type WrapperProps = { bgColor: ButtonBgColor }

export const Wrapper = styled.button<WrapperProps>`
  ${({ theme, bgColor }) => css`
    cursor: pointer;
    padding: 0.6rem 2.7rem;
    font-weight: ${theme.font.bold};
    font-size: ${theme.font.sizes.medium};
    border-radius: ${theme.border.radius};
    background-color: ${theme.colors[bgColor]};
    color: ${bgColor === 'white' ? theme.colors.black : theme.colors.white};
    transition: ${theme.transition.default};

    &:hover {
      color: ${lighten(0.1, theme.colors[bgColor])};
    }

    &:disabled {
      cursor: not-allowed;
      filter: saturate(30%);
    }

    ${bgColor === 'white'
      ? css`
          border: 1px solid ${theme.colors.gray};
          &:hover {
            color: ${darken(0.5, theme.colors[bgColor])};
          }
        `
      : css`
          border: none;
          &:hover {
            color: ${lighten(0.1, theme.colors[bgColor])};
          }
        `}
  `}
`
