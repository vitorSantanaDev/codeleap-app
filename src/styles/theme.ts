export default {
  grid: {
    container: '130rem'
  },
  border: {
    radius: '0.8rem'
  },
  font: {
    family:
      "Poppins, -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Oxygen, Ubuntu, Cantarell, 'Open Sans', 'Helvetica Neue', sans-serif",
    normal: 400,
    bold: 700,
    sizes: {
      xsmall: '1.2rem',
      small: '1.4rem',
      medium: '1.6rem',
      large: '1.8rem',
      xlarge: '2.0rem',
      xxlarge: '2.8rem',
      huge: '5.2rem'
    }
  },
  colors: {
    primary: '#7695EC',
    white: '#FFFFFF',
    black: '#000000',
    gray: '#999999',
    dark_gray: '#777777',
    light_gray: '#CCCCCC',
    red: '#FF5151',
    green: '#47B960',
    overlay_color: 'rgba(119, 119, 119, 0.8)'
  },
  spacings: {
    xxsmall: '0.8rem',
    xsmall: '1.6rem',
    small: '2.4rem',
    medium: '3.2rem',
    large: '4.0rem',
    xlarge: '4.8rem',
    xxlarge: '5.6rem'
  },
  layers: {
    base: 10,
    menu: 20,
    overlay: 30,
    modal: 40,
    alwaysOnTop: 50
  },
  transition: {
    fast: '0.1s ease-in-out',
    default: '0.3s ease-in-out'
  }
} as const
