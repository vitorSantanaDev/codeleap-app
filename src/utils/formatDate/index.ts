import {
  format,
  parseISO,
  minutesToHours,
  secondsToMinutes,
  differenceInSeconds
} from 'date-fns'

export const formattingStringWithTheTimePassedBetweenTwoDate = (
  hours: number,
  minutes: number,
  seconds: number
): string => {
  if (hours < 1 && minutes < 1) {
    return `${seconds} seconds ago`
  }

  let formattedTime = ''

  if (hours > 0) {
    formattedTime = `${hours} ${hours > 1 ? 'hours' : 'hour'} ago`
  } else if (hours < 1) {
    formattedTime = `${minutes} minutes ago`
  }

  return formattedTime
}

export const rangeOfTimePassedBetweenTwoDates = (
  start: string,
  end: string
): string => {
  const startDate = parseISO(start)
  const endDate = parseISO(end)

  const allSeconds = differenceInSeconds(endDate, startDate)
  const allMinutes = secondsToMinutes(allSeconds)

  const minutesInSeconds = allMinutes * 60
  const seconds = allSeconds - minutesInSeconds

  const hours = minutesToHours(allMinutes)

  const hoursInMinutes = hours * 60

  const minutes = allMinutes - hoursInMinutes

  return hours > 24
    ? format(new Date(startDate), 'dd/MM/yyyy')
    : formattingStringWithTheTimePassedBetweenTwoDate(hours, minutes, seconds)
}
