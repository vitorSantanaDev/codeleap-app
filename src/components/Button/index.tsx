import React from 'react'

import { ButtonProps } from './types'

import * as S from './styles'

const Button: React.FC<ButtonProps> = ({
  children,
  bgColor = 'primary',
  ...restProps
}) => {
  return (
    <S.Wrapper bgColor={bgColor} {...restProps}>
      {children}
    </S.Wrapper>
  )
}

export default Button
