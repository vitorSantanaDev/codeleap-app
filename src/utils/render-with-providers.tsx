import { Provider } from 'react-redux'
import React, { PropsWithChildren } from 'react'

import theme from 'styles/theme'
import { ThemeProvider } from 'styled-components'
import { RenderOptions, render } from '@testing-library/react'

import { RootState } from 'redux/store'
import { userReducer } from 'redux/actions/user-slice'
import { PreloadedState, configureStore } from '@reduxjs/toolkit'
import { QueryClient, QueryClientProvider } from 'react-query'

interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  initialState?: PreloadedState<RootState>
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  store?: any
}

export const renderWithProviders = (
  ui: React.ReactElement,
  {
    initialState = {},
    store = configureStore({
      reducer: { user: userReducer },
      preloadedState: initialState
    }),
    ...renderOptions
  }: ExtendedRenderOptions = {}
) => {
  function Wrapper({ children }: PropsWithChildren<unknown>): JSX.Element {
    return (
      <QueryClientProvider client={new QueryClient()}>
        <Provider store={store}>
          <ThemeProvider theme={theme}>{children}</ThemeProvider>
        </Provider>
      </QueryClientProvider>
    )
  }

  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) }
}
