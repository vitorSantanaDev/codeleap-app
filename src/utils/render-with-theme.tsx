import React, { ReactElement } from 'react'

import theme from 'styles/theme'
import { render } from '@testing-library/react'
import { ThemeProvider } from 'styled-components'

export const renderWithTheme = (component: ReactElement) =>
  render(<ThemeProvider theme={theme}>{component}</ThemeProvider>)
