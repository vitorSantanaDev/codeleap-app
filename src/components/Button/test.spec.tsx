import React from 'react'
import { renderWithProviders } from 'utils/render-with-providers'
import Button from '.'
import { screen } from '@testing-library/react'
import theme from 'styles/theme'

describe('<Button />', () => {
  it('should render button background with default color', () => {
    renderWithProviders(<Button>Button</Button>)

    const button = screen.getByRole('button', { name: /button/i })
    expect(button).toBeInTheDocument()

    expect(button).toHaveStyle({ backgroundColor: theme.colors.primary })
  })

  it('should render button with background red', () => {
    renderWithProviders(<Button bgColor="red">Button</Button>)

    const button = screen.getByRole('button', { name: /button/i })

    expect(button).toHaveStyle({
      backgroundColor: theme.colors.red,
      color: theme.colors.white
    })
  })

  it('should render button with background red', () => {
    renderWithProviders(<Button bgColor="red">Button</Button>)

    const button = screen.getByRole('button', { name: /button/i })

    expect(button).toHaveStyle({
      backgroundColor: theme.colors.red,
      color: theme.colors.white
    })
  })

  it('should render button with background green', () => {
    renderWithProviders(<Button bgColor="green">Button</Button>)

    const button = screen.getByRole('button', { name: /button/i })

    expect(button).toHaveStyle({
      backgroundColor: theme.colors.green,
      color: theme.colors.white
    })
  })

  it('should render button with background white, and borders with color gray', () => {
    renderWithProviders(<Button bgColor="white">Button</Button>)

    const button = screen.getByRole('button', { name: /button/i })

    expect(button).toHaveStyle({
      backgroundColor: theme.colors.white,
      color: theme.colors.black,
      border: `1px solid ${theme.colors.gray}`
    })
  })

  it('should render button with disabled style', () => {
    renderWithProviders(<Button bgColor="white">Button</Button>)

    const button = screen.getByRole('button', { name: /button/i })

    expect(button).toHaveStyleRule('cursor', 'not-allowed', {
      modifier: ':disabled'
    })
  })

  it('should match snapshot', () => {
    const { container } = renderWithProviders(<Button>Button</Button>)
    expect(container.firstChild).toMatchSnapshot()
  })
})
