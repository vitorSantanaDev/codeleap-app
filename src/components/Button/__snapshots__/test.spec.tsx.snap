// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> should match snapshot 1`] = `
.c0 {
  cursor: pointer;
  padding: 0.6rem 2.7rem;
  font-weight: 700;
  font-size: 1.6rem;
  border-radius: 0.8rem;
  background-color: #7695EC;
  color: #FFFFFF;
  -webkit-transition: 0.3s ease-in-out;
  transition: 0.3s ease-in-out;
  border: none;
}

.c0:hover {
  color: #a3b8f2;
}

.c0:disabled {
  cursor: not-allowed;
  -webkit-filter: saturate(30%);
  filter: saturate(30%);
}

.c0:hover {
  color: #a3b8f2;
}

<button
  class="c0"
>
  Button
</button>
`;
